src/analysis.rs

```diff
@@ -27,358 +23,832 @@ pub fn generate_blackman(size: usize) -> impl Iterator<Item = f32> {
+        // get real and imaginary frequency data.
         r2c.process_with_scratch(input, output, scratch).unwrap();
 
-        // smoothing over time
-        previous_block
+        // Notes from chromium source code (tbc)
+        //
+        // cf. third_party/blink/renderer/platform/audio/fft_frame.h"
+        // ```
+        // Since x[n] is assumed to be real, X[k] has complex conjugate symmetry with
+        // X[N-k] = conj(X[k]).  Thus, we only need to keep X[k] for k = 0 to N/2.
+        // But since X[0] is purely real and X[N/2] is also purely real, so we could
+        // place the real part of X[N/2] in the imaginary part of X[0].  Thus
+        // for k = 1 to N/2:
+        //
+        //   real_data[k] = Re(X[k])
+        //   imag_data[k] = Im(X[k])
+        //
+        // and
+        //
+        //   real_data[0] = Re(X[0]);
+        //   imag_data[0] = Re(X[N/2])
+        // ```
+        //
+        // It seems to be why their FFT return only `fft_size / 2` components
+        // instead `fft_size * 2 + 1`, they pack DC and Nyquist bins together.
+        //
+        // However in their `realtime_analyser` they then remove the packed nyquist
+        // imaginary component:
+        // cf. third_party/blink/renderer/modules/webaudio/realtime_analyser.h
+        // ```
+        // // Blow away the packed nyquist component.
+        // imag[0] = 0;
+        // ```
+        // In our case, it seems we can thus just ignore the Nyquist information
+        // and take the DC bin as it is
+
+        let normalize_factor = 1. / fft_size as f32;
+
+        last_fft_output
             .iter_mut()
             .zip(output.iter())
-            .for_each(|(p, c)| {
-                *p = smoothing_time_constant * *p + (1. - smoothing_time_constant) * c.norm()
+            .for_each(|(o, c)| {
+                let norm = c.norm() * normalize_factor;
+                let value = smoothing_time_constant * *o + (1. - smoothing_time_constant) * norm;
+                *o = if value.is_finite() { value } else { 0. };
+            });
+    }
+
+    pub fn get_float_frequency_data(&mut self, dst: &mut [f32], current_time: f64) {
+        let frequency_bin_count = self.frequency_bin_count();
+
+        // [spec] If another call to getByteFrequencyData() or getFloatFrequencyData()
+        // occurs within the same render quantum as a previous call, the current
+        // frequency data is not updated with the same data. Instead, the previously
+        // computed data is returned.
+        if current_time != self.last_fft_time {
+            self.compute_fft();
+            self.last_fft_time = current_time;
+        }
+
+        // [spec] Write the current frequency data into array. If array’s byte
+        // length is less than frequencyBinCount, the excess elements will be
+        // dropped. If array’s byte length is greater than the frequencyBinCount
+        let len = dst.len().min(frequency_bin_count);
```

Comment:
`zip` will stop when the end of either buffer is reached, so I don't think `take` is necessary

Reply:
actually the `last_fft_output` buffer is never resized, so we need the `take` here

Reply:
good point